README.md

## The `speech` package <img src='man/figures/logo.png' align="right" height="180" />

*Nicolás Schmidt, Diego Luján, Juan Andrés Moraes*

<!-- badges: start -->

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version-ago/speech)](https://cran.r-project.org/package=speech)
[![](https://img.shields.io/badge/devel%20version-0.1.6-orange.svg)](https://github.com/Nicolas-Schmidt/speech)
[![R build
status](https://github.com/Nicolas-Schmidt/speech/workflows/R-CMD-check/badge.svg)](https://github.com/Nicolas-Schmidt/speech/actions)
[![R-CMD-check](https://github.com/Nicolas-Schmidt/speech/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/Nicolas-Schmidt/speech/actions/workflows/R-CMD-check.yaml)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/speech?color=blue)](https://www.r-pkg.org/pkg/speech)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/speech?color=blue)](https://www.r-pkg.org/pkg/speech)
<!-- badges: end -->

### Description

Converts the floor speeches of Uruguayan legislators, extracted from the
parliamentary minutes, to tidy data.frame where each observation is the
intervention of a single legislator.

### Installation

``` r
# Install speech from CRAN
install.packages("speech")

# The development version from GitHub:
if (!require("remotes")) install.packages("remotes")
remotes::install_github("Nicolas-Schmidt/speech")
```

### Data generation process

#### 1 - Floor speeches

<img src="man/figures/fspeech.png" style="margin-center:30px;margin-bottom:5px;" width="600" align="center"></a>

#### 2 - Data extraction

<img src="man/figures/fspeechextract.png" style="margin-center:30px;margin-bottom:5px;" width="600" align="center"></a>

#### 3 - First construction of the data set: `speech::speech_build()`

<img src="man/figures/fspeechnocompiler.png" style="margin-center:30px;margin-bottom:5px;" width="900" align="center"></a>

#### 4 - Final data set: `speech::speech_build(., compiler = TRUE)`

<img src="man/figures/fspeechcompiler.png" style="margin-center:30px;margin-bottom:5px;" width="900" align="center"></a>

### Example

You can see more complex examples in the following
[**link**](https://github.com/Nicolas-Schmidt/speech-ejemplos).

``` r
library(speech)
url <- "https://parlamento.gub.uy/documentosyleyes/documentos/diarios-de-sesion/6084/IMG"
text <- speech::speech_build(file = url)
text
#> # A tibble: 24 × 7
#>    legislator speech                  chamber date       legislature id      sex
#>    <chr>      <chr>                   <chr>   <date>           <int> <chr> <dbl>
#>  1 BORDABERRY SEÑOR BORDABERRY. Pido… COMISI… 2019-09-17          48 2019…     1
#>  2 BORDABERRY SEÑOR BORDABERRY. Prop… COMISI… 2019-09-17          48 2019…     1
#>  3 AVIAGA     SEÑORA AVIAGA. Pido la… COMISI… 2019-09-17          48 2019…     0
#>  4 AVIAGA     SEÑORA AVIAGA. En el m… COMISI… 2019-09-17          48 2019…     0
#>  5 GOÑI       SEÑOR GOÑI. Pido la pa… COMISI… 2019-09-17          48 2019…     1
#>  6 GOÑI       SEÑOR GOÑI. El Frente … COMISI… 2019-09-17          48 2019…     1
#>  7 MAHIA      SEÑOR MAHIA. Pido la p… COMISI… 2019-09-17          48 2019…     1
#>  8 MAHIA      SEÑOR MAHIA. Gracias, … COMISI… 2019-09-17          48 2019…     1
#>  9 ABDALA     SEÑOR ABDALA. Voto por… COMISI… 2019-09-17          48 2019…     1
#> 10 ASTI       SEÑOR ASTI. Obviamente… COMISI… 2019-09-17          48 2019…     1
#> # ℹ 14 more rows


speech_check(text, initial = c("A", "M"))
#> $A
#>   legislator
#> 1     ABDALA
#> 2       ASTI
#> 3     AVIAGA
#> 
#> $M
#>   legislator
#> 1      MAHIA
#> 2     MERONI


text <- speech::speech_build(file = url, compiler = TRUE)
text
#> # A tibble: 11 × 7
#>    legislator legislature chamber             date       id         speech   sex
#>    <chr>            <int> <chr>               <date>     <chr>      <chr>  <dbl>
#>  1 ABDALA              48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1
#>  2 ASTI                48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1
#>  3 AVIAGA              48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     0
#>  4 BORDABERRY          48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1
#>  5 GOÑI                48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1
#>  6 LAZO                48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     0
#>  7 MAHIA               48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1
#>  8 MERONI              48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1
#>  9 PEREYRA             48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     0
#> 10 TOURNE              48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     0
#> 11 VIERA               48 COMISION PERMANENTE 2019-09-17 20190917c… SEÑOR…     1


text$word <- speech_word_count(text$speech)

dplyr::glimpse(text)
#> Rows: 11
#> Columns: 8
#> $ legislator  <chr> "ABDALA", "ASTI", "AVIAGA", "BORDABERRY", "GOÑI", "LAZO", …
#> $ legislature <int> 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48
#> $ chamber     <chr> "COMISION PERMANENTE", "COMISION PERMANENTE", "COMISION PE…
#> $ date        <date> 2019-09-17, 2019-09-17, 2019-09-17, 2019-09-17, 2019-09-17…
#> $ id          <chr> "20190917c0001", "20190917c0001", "20190917c0001", "20190…
#> $ speech      <chr> "SEÑOR ABDALA. Voto por la señora legisladora Daisy Tourné…
#> $ sex         <dbl> 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1
#> $ word        <int> 398, 46, 105, 951, 98, 103, 126, 12, 12, 111, 8
```

### Possible application

``` r
library(magrittr)

minchar <- function(string, min = 3){
    string <- stringr::str_remove_all(string, "[[:punct:]]")
    string <- unlist(strsplit(string, " "))
    string[nchar(string) > min]
}

text$speech %>% 
    minchar(., min = 4) %>%  
    quanteda::corpus() %>% 
    quanteda::dfm(remove = c("señor", "señora")) %>% 
    quanteda.textplots::textplot_wordcloud(color = rev(RColorBrewer::brewer.pal(10, "RdBu")))
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

``` r
library(ggplot2)
#> Warning: package 'ggplot2' was built under R version 4.2.3

text$speech %>% 
    minchar(., min = 4) %>%  
    tibble::enframe() %>% 
    tidytext::unnest_tokens(word, value) %>%
    dplyr::count(word, sort = TRUE) %>%
    dplyr::mutate(word = stats::reorder(word, n)) %>%
    dplyr::filter(!stringr::str_detect(word, "^señor")) %>% 
    .[1:40,] %>% 
    ggplot(aes(word, n)) +
        geom_col(col = "black", fill = "#00A08A", width = .7) +
        labs(x = "", y = "") +
        coord_flip() +
        theme_minimal()
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="80%" />

### Detecting roll-call votes in parliamentary speeches

``` r
urls <- speech_url(chamber = "D", days = c("2002-06-12", "2004-04-14"))
rollcall <- speech_rollcall(file =  urls)

rollcall
#> # A tibble: 165 × 10
#>    legislator       vote argument speech chamber date       legislature rollcall
#>    <chr>           <dbl>    <dbl> <chr>  <chr>   <date>           <int>    <int>
#>  1 ABDALA              0        0 SEÑOR… CAMARA… 2002-06-12          45        1
#>  2 AGAZZI              1        1 SEÑOR… CAMARA… 2002-06-12          45        1
#>  3 AMEN VAGHETTI       0        0 SEÑOR… CAMARA… 2002-06-12          45        1
#>  4 AMORIN BATLLE       0        0 SEÑOR… CAMARA… 2002-06-12          45        1
#>  5 ARAUJO              0        1 SEÑOR… CAMARA… 2002-06-12          45        1
#>  6 ARGENZIO            0        1 SEÑOR… CAMARA… 2002-06-12          45        1
#>  7 ARGIMON             0        0 SEÑOR… CAMARA… 2002-06-12          45        1
#>  8 ARRARTE FERNAN…     1        0 SEÑOR… CAMARA… 2002-06-12          45        1
#>  9 ARREGUI             1        1 SEÑOR… CAMARA… 2002-06-12          45        1
#> 10 BARAIBAR            1        0 SEÑOR… CAMARA… 2002-06-12          45        1
#> # ℹ 155 more rows
#> # ℹ 2 more variables: id <chr>, sex <dbl>

summary(rollcall)
#> # A tibble: 2 × 10
#>   Chamber Date       Legislators Affirmative Negative prop_AF prop_NG prop_women
#>   <chr>   <date>           <int>       <dbl>    <dbl>   <dbl>   <dbl>      <dbl>
#> 1 CRR     2002-06-12          92          45       47    48.9    51.1       13.0
#> 2 CRR     2004-04-14          73          32       41    43.8    56.2       15.1
#> # ℹ 2 more variables: prop_arg <dbl>, rc <int>
```

#### Citation

To cite package`speech` in publications, please use:

``` r
citation(package = 'speech')
#> 
#> To cite speech in publications use:
#> 
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {Estimating Parties’ Policy Positions in Uruguay: Comparing Scaling Methods Based on Legislative Speeches and Roll-Call Votes},
#>     author = {Diego Luján and Nicolás Schmidt and Juan A. Moraes},
#>     journal = {Latin American Politics and Society},
#>     year = {2023},
#>     volume = {0},
#>     number = {0},
#>     pages = {1-17},
#>     url = {doi:10.1017/lap.2023.12},
#>   }
```

#### Maintainer

Nicolas Schmidt (<nschmidt@cienciassociales.edu.uy>)


NEWS.md
# speech 0.1.6

* The change in this version is not visible to the user but it is important since otherwise the url assignment in the functions would give errors. This is because the parliament website incorporated a pdf viewer on the page.


# speech 0.1.5

* A function is added (`speech_rollcall()`) to detect roll-call votes from parliamentary speeches.
* The argument `days` is added to the function `speech_url()`


# speech 0.1.4

* {tabulizer} functions are replaced by {tm} functions.


# speech 0.1.3

 * add `speech_url()`
 * add `speech_uncompiler()`

# speech 0.1.2

* variable sex added in output function `speech_build()`
* argument `rm.name` in `speech_word_count()` added

# speech 0.1.1

* change `dplyr::bind_rows()` for `do.call(rbind, .)`
* change `dplyr::mutate()` for `transform()`
* typeof legislature variable now is numeric in `speech::speech_recompiler()`
* typeof date variable now is date
* The value of the `compiler_by` argument in the `speech_recompailer()` function is modified. 
Before: `c("legislator", "legislature", "chamber", "party")`, now: `c("legislator", "legislature", "chamber", "date", "id")`


# speech 0.1.0

* CRAN-RELEASE


R/speech_build.R
#' @title Transform speeches in pdf to data.frame
#' @description It allows to extract the individual speeches of each legislator
#'     in a document and obtain a data.frame.
#' @param file list or character vector specifying the path or URL to a PDF file.
#'     It can be one or more files.
#' @param add.error.sir character vector. It allows to specify different ways in which
#'     the term that orders the speeches could be miswritten: sir. By default it is \code{NULL}.
#' @param rm.error.leg character vector. It allows to add legislator's names
#'     to be eliminated. By default it is \code{NULL}.
#'     By default, "PRESIDENTE", "SECRETARIO", "SUBSECRETARIO", and "MINISTRO" are eliminated.
#' @param compiler logical. When the checking of the process of conversion from pdf to data frame
#'     is completed, it is necessary to compile the data frame. To compile implies to unite all the
#'     speeches of each of the legislators for each document. As it is an operation
#'     that must be carried out after making corrections, it is necessary to opt for it.
#'     By default it is \code{FALSE}.
#' @param quality logical. If \code{TRUE}, two quality indicators are added about
#'     the process, according to the quality of the document.
#'     \itemize{
#'             \item{index_1: Proportion of the text recovered according to the original document
#'             (\code{param = list(char = 6500, drop.page = 2)}) that must have the document.}
#'             \item{index_2: Proportion of the final text as a function of the
#'             recovered text. It is the proportion of the document in which there are
#'             only interventions by legislators.}
#'            }
#' @param param list of length 2 with magnitudes for arguments "character for page" and "drop page
#'     non evaluate" respectively. The default values are the median characters of
#'     8500 documents that make up the speech datasets.
#' @details This function converts PDF documents to data.frame. The conversion is
#'     made by seeking interventions of legislators from the word "SENOR". As the
#'     quality of PDF files is not always the best it is recommended to verify that
#'     no legislator is omitted in the data.frame construction process. To make
#'     corrections of the word "SENOR" is that the argument \code{add.error.sir}
#'     should be used. The function has a long list of different ways in which
#'     the word "SENOR" may be written in a document, but not all possible future
#'     problems are covered. When the PDF document is a scan that was treated with
#'     an OCR, it should be checked with greater caution to ensure that the operation
#'     was performed correctly.
#' @return data.frame class \code{puy} with the following variables:
#'      \itemize{
#'             \item{\code{legislator}: name of the legislators}
#'             \item{\code{speech}: speeches by legislators}
#'             \item{\code{date}: session date}
#'             \item{\code{id}: name \code{file}}
#'             \item{\code{legislature}: legislature id (period of government)}
#'             \item{\code{sex}: sex}
#'             \item{\code{chamber}: chamber to which the document belongs.
#'             It can be: Chamber of Representatives, Senate, General Assembly or Permanent Commission.}
#'            }
#'          If quality is TRUE, the following are added:
#'      \itemize{
#'             \item{\code{index_1}: index_1}
#'             \item{\code{index_2}: index_2}
#'            }
#'
#' @examples
#' \donttest{
#' # url <- speech::speech_url(chamber = "C", from = "17-09-2019", to = "17-09-2019")
#' # out <- speech_build(file = url)
#'
#' # out <- speech_build(file = url, compiler = FALSE,
#' #                     quality = TRUE,
#' #                     add.error.sir = c("SEf'IOR"),
#' #                     rm.error.leg = c("PRtSIDENTE", "SUB", "PRfSlENTE"),
#' #                     param = list(char = 6000, drop.page = 3))
#'
#' # out <- list.files(pattern = "*.pdf") %>% speech_build()
#'
#' # out <- list.files(pattern = "*.pdf") %>%
#' #     speech_build(., compiler = TRUE, param = list(char = 4500, drop.page = 3))
#' }
#' @export

speech_build <- function(file, add.error.sir = NULL, rm.error.leg = NULL, compiler = FALSE,
                   quality = FALSE, param = list(char = 6500, drop.page = 2)){

        file <- sfile(file)
        out  <-
                file %>%
                purrr::map(speech.pow,
                           add.error.sir = add.error.sir,
                           rm.error.leg = rm.error.leg,
                           quality = quality,
                           param = list(char = param$char, drop.page = param$drop.page)
                )

        out <- out[!unlist(lapply(out, is.null))]
        out <- out[purrr::map_int(out, nrow) != 0]
        out <- do.call(rbind, out)
        out$speech <- stringr::str_squish(out$speech)

        if(compiler){
            compiler(tidy_speech = out,
                     compiler_by = c("legislator",
                                     "legislature",
                                     "chamber",
                                     "date",
                                     "id"))
        } else {
            invisible(
                    SError(
                            add_sex(
                                    clean_t(
                                            out))))
        }
}




